### ISSUE TITLE
MacCatalyst clang: error: linker command failed with exit code 1 

### ISSUE THREAD

**Issue by cdavidyoung** [xamarin/xamarin-macios#19258]
### Steps to Reproduce

1. Sorry that I don't have an easy way to reproduce this.  I tried a new Maui Hello World project and it builds and runs fine for MacCatalyst. @rolfbjarne asked me to create a new issue here.
2. My project previously built and ran fine for MacCatalyst.  It was only after that last VS updates that I started getting the clang error.
3. Here is the error I am getting:

/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3): error : clang++ exited with code 1:
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3): error : ld: warning: ignoring duplicate libraries: '-lSystem.IO.Compression.Native', '-lSystem.Native', '-lSystem.Net.Security.Native', '-lSystem.Security.Cryptography.Native.Apple', '-lcompression', '-lmonosgen-2.0', '-lxamarin-dotnet-debug'
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3): error : ld: building for 'macCatalyst', but linking in object file (/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame[x86_64][2](VbrTag.o)) built for 'iOS-simulator'
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3): error : clang: error: linker command failed with exit code 1 (use -v to see invocation)
    0 Warning(s)
    1 Error(s)

### Expected Behavior

Link successfully

### Actual Behavior

clang: error: linker command failed with exit code 1 

### Environment

<details>
<summary>Version information</summary>

```
Visual Studio Community 2022 for Mac
Version 17.6.5 (build 417)
Installation UUID: 9ecc4e19-b627-40f2-ab6d-f56425c5174e

Runtime
.NET 7.0.3 (64-bit)
Architecture: Arm64
Microsoft.macOS.Sdk 13.1.1007; git-rev-head:8afca776a0a96613dfb7200e0917bb57f9ed5583; git-branch:release/7.0.1xx-xcode14.2

Roslyn (Language Service)
4.6.0-3.23180.6+99e956e42697a6dd886d1e12478ea2b27cceacfa

NuGet
Version: 6.4.0.117

.NET SDK (Arm64)
SDK: /usr/local/share/dotnet/sdk/7.0.309/Sdks
SDK Versions:
	7.0.309
	7.0.308
	7.0.302
	7.0.203
	7.0.202
	7.0.201
	7.0.200
	7.0.103
MSBuild SDKs: /Applications/Visual Studio.app/Contents/MonoBundle/MSBuild/Current/bin/Sdks

.NET Runtime (Arm64)
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	7.0.12
	7.0.11
	7.0.5
	7.0.4
	7.0.3

Xamarin.Profiler
Version: 1.8.0.49
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Apple Developer Tools
Xcode: 14.3 21812
Build: 15A240d

Xamarin.Mac
Version: 9.3.0.18 Visual Studio Community
Hash: 9d266025e
Branch: xcode14.3
Build date: 2023-09-06 19:52:26-0400

Xamarin.Android
Version: 13.2.2.0 (Visual Studio Community)
Commit: xamarin-android/d17-5/45b0e14
Android SDK: /Users/charlesyoung/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		8.0  (API level 26)
		12.0 (API level 31)
		8.1  (API level 27)
		11.0 (API level 30)
		10.0 (API level 29)
		13.0 (API level 33)

SDK Command-line Tools Version: 7.0
SDK Platform Tools Version: 33.0.2
SDK Build Tools Version: 33.0.0

Build Information: 
Mono: d9a6e87
Java.Interop: xamarin/java.interop/d17-5@149d70fe
SQLite: xamarin/sqlite/3.40.1@68c69d8
Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-5@ca1552d

Microsoft Build of OpenJDK
Java SDK: /Library/Java/JavaVirtualMachines/microsoft-11.jdk
11.0.16.1
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Eclipse Temurin JDK
Java SDK: /Library/Java/JavaVirtualMachines/temurin-8.jdk
1.8.0.302
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 17.6.0.50
Hash: a715dca
Branch: HEAD
Build date: 2023-10-04 18:09:18 UTC

Android Device Manager
Version: 0.0.0.1309
Hash: 06e3e77
Branch: HEAD
Build date: 2023-10-04 18:09:18 UTC

Xamarin Designer
Version: 17.6.3.9
Hash: 2648399ae8
Branch: remotes/origin/d17-6
Build date: 2023-10-04 18:09:13 UTC

Xamarin.iOS
Version: 16.4.0.18 Visual Studio Community
Hash: 9d266025e
Branch: xcode14.3
Build date: 2023-09-06 19:52:27-0400

Build Information
Release ID: 1706050417
Git revision: 6d6585a706becbd4a5be3b0e99ace260dfdf5748
Build date: 2023-10-04 18:07:26+00
Build branch: release-17.6
Build lane: release-17.6

Operating System
Mac OS X 14.0.0
Darwin 23.0.0 Darwin Kernel Version 23.0.0
    Fri Sep 15 14:41:34 PDT 2023
    root:xnu-10002.1.13~1/RELEASE_ARM64_T8103 arm64

Enabled user installed extensions
Project System Tools 0.13


```

</details>

##
[jiffyLog_Debug_Build_2023-10-17T12_54_04.9444050Z.msbuild.log](https://github.com/xamarin/xamarin-macios/files/13032910/jiffyLog_Debug_Build_2023-10-17T12_54_04.9444050Z.msbuild.log)
# Example Project (If Possible)

**Comment by rolfbjarne** [comment]
Can you try adding this to your project to see if it works?

```xml
<ItemGroup>
    <MtouchExtraArgs>$(MtouchExtraArgs) -gcc_flags "-Xlinker -ld_classic"</MtouchExtraArgs>
    <_MainLinkerFlags Include="-ld_classic" />
</ItemGroup>
```

**Comment by cdavidyoung** [comment]
@Rolf Luetcke ***@***.***> Thanks for trying to help.  These 4
lines added to my .csproj file just like other <ItemGroup> lines results in
the project not loading.

**Comment by cdavidyoung** [comment]
@Rolf Luetcke ***@***.***>

[image: image.png]
I now get the same error clang.

**Comment by rolfbjarne** [comment]
Can you get an updated binlog?

**Comment by cdavidyoung** [comment]
@Rolf Luetcke ***@***.***> Here is updated binlog.

**Comment by inderpalaulakh** [comment]
I am seeing same issue. The issue exists only on iPhone, Simulators works fine. 
Meanwhile the issue is not fixed, please try Simulator.

**Comment by cdavidyoung** [comment]
Just to be clear, I have no problem building for iPhone.  This is just a
MacCatalyst issue for me.

**Comment by rolfbjarne** [comment]
@cdavidyoung the build log didn't attach properly, also please get the binary version as explained here: https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#binary-build-logs

**Comment by cdavidyoung** [comment]
@rolfbjarne Thanks for your continued support.  This is very confusing.  The version of VS on my Mac is 17.6.5, which is the "Visual Studio 2022 for Mac" latest and greatest, I believe.  However, the web page for the build log says "Support for design-time build binary logs should be available in Visual Studio for Mac 8.5", which is Visual Studio 2019.  The "Open Binary Log File" option in my VS is greyed out and not available.

Which version of VS for Mac should I be using?

**Comment by cdavidyoung** [comment]
@rolfbjarne  Should I uninstall Visual Studio 2022 for Mac and install Visual Studio 2019 to try and get the binary log?

**Comment by rolfbjarne** [comment]
> The "Open Binary Log File" option in my VS is greyed out and not available.

That's weird, I've never seen that before.

In any case, it might be easier to do this from the command line:

Try opening the Terminal app, and execute:

```shell
cd /Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj
dotnet build /bl:msbuild.binlog /f:net7.0-maccatalyst jiffyLog.csproj
```

and then zip up and attach the msbuild.binlog file that was created.

**Comment by cdavidyoung** [comment]
This is the result when I try the command line build.  BTW, it always says
that workload updates are available and when I run the update it comes back
and says "Successfully updated workload(s): maui-maccatalyst maui-ios
maui-android".

***@***.*** jiffyLog2 % dotnet build /bl:msbuild.binlog
/f:net7.0-maccatalyst jiffyLog.csproj
MSBuild version 17.6.10+2679cf5a9 for .NET
MSBUILD : error MSB1001: Unknown switch.
    Full command line: '/usr/local/share/dotnet/sdk/7.0.309/MSBuild.dll
-maxcpucount -verbosity:m -restore -consoleloggerparameters:Summary
/bl:msbuild.binlog /f:net7.0-maccatalyst jiffyLog.csproj
-distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/usr/local/share/dotnet/sdk/7.0.309/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/usr/local/share/dotnet/sdk/7.0.309/dotnet.dll'
  Switches appended by response files:
Switch: /f:net7.0-maccatalyst

For switch syntax, type "MSBuild -help"

Workload updates are available. Run `dotnet workload list` for more
information.
***@***.*** jiffyLog2 %

**Comment by rolfbjarne** [comment]
Can you try this and see if it works?

> dotnet build /bl:msbuild.binlog

**Comment by cdavidyoung** [comment]
This seemed to work and I attached the binlog.

***@***.*** jiffyLog2 % dotnet build /bl:msbuild.binlog
MSBuild version 17.6.10+2679cf5a9 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  Detected signing identity:

    Code Signing Key: "Apple Distribution: Charles Young (83Q3M48Y8B)"
(ED3D7B75076148A331F7BAE7619953BE8DB5B0CF)
    Provisioning Profile: "CacheAll CacheAll Mac App Store"
(9967907e-e767-4d6b-a577-e5412e7120da)
    Bundle Id: com.CacheAll
    App Id: 83Q3M48Y8B.com.CacheAll
  jiffyLog ->
/Users/charlesyoung/Projects/jiffyLog2/bin/Debug/net7.0-maccatalyst/maccatalyst-x64/CacheAll.dll
  Detected signing identity:

    Bundle Id: com.CacheAll
    App Id: com.CacheAll
  Optimizing assemblies for size may change the behavior of the app. Be
sure to test after publishing. See: https://aka.ms/dotnet-illink
  jiffyLog ->
/Users/charlesyoung/Projects/jiffyLog2/bin/Debug/net7.0-ios/iossimulator-x64/CacheAll.dll
  Optimizing assemblies for size may change the behavior of the app. Be
sure to test after publishing. See: https://aka.ms/dotnet-illink
/Users/charlesyoung/.nuget/packages/instances/3.0.0/contentFiles/any/netstandard2.0/README.md
: warning XA0101: @(Content) build action is not supported
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-android]
  jiffyLog ->
/Users/charlesyoung/Projects/jiffyLog2/bin/Debug/net7.0-android/CacheAll.dll
  Tool xcrun execution finished (exit code = 1).

  ld: in
/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame(VbrTag.o),
building for Mac Catalyst, but linking in object file built for iOS
Simulator, file
'/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame'
for architecture x86_64
  clang: error: linker command failed with exit code 1 (use -v to see
invocation)

/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : clang++ exited with code 1:
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : ld: in
/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame(VbrTag.o),
building for Mac Catalyst, but linking in object file built for iOS
Simulator, file
'/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame'
for architecture x86_64
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : clang: error: linker command failed with exit code 1 (use -v to see
invocation)
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]

Build FAILED.

/Users/charlesyoung/.nuget/packages/instances/3.0.0/contentFiles/any/netstandard2.0/README.md
: warning XA0101: @(Content) build action is not supported
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-android]
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : clang++ exited with code 1:
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : ld: in
/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame(VbrTag.o),
building for Mac Catalyst, but linking in object file built for iOS
Simulator, file
'/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame'
for architecture x86_64
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.4.7107/targets/Xamarin.Shared.Sdk.targets(1272,3):
error : clang: error: linker command failed with exit code 1 (use -v to see
invocation)
[/Users/charlesyoung/Projects/jiffyLog2/jiffyLog.csproj::TargetFramework=net7.0-maccatalyst]
    1 Warning(s)
    1 Error(s)

Time Elapsed 00:00:26.51

Workload updates are available. Run `dotnet workload list` for more
information.
***@***.*** jiffyLog2 %

**Comment by rolfbjarne** [comment]
Looks like attachments through emails aren't attached to the issue, so I can't access it. I'm believe you'll have to attach using the web UI.

**Comment by cdavidyoung** [comment]
[msbuild.binlog.zip](https://github.com/xamarin/xamarin-macios/files/13069991/msbuild.binlog.zip)

**Comment by rolfbjarne** [comment]
That binlog doesn't shows a build that doesn't do much, because it's an incremental build (everything's already been done).

Can you first change the MtouchExtraArgs to the csproj to be like this (adding `-vvvv` for increased verbosity):

```xml
<MtouchExtraArgs>$(MtouchExtraArgs) -gcc_flags "-Xlinker -ld_classic" -vvvv</MtouchExtraArgs>
```

Then clean your project directory:

```
dotnet clean
```

before building again and finally uploading the resulting msbuild.binlog?

**Comment by cdavidyoung** [comment]
[msbuild.binlog.zip](https://github.com/xamarin/xamarin-macios/files/13073106/msbuild.binlog.zip)

**Comment by cdavidyoung** [comment]
@rolfbjarne Did you get my last binlog?

**Comment by rolfbjarne** [comment]
> @rolfbjarne Did you get my last binlog?

Yes, I did, and there's something suspicious here:

> ILLink: Ignoring unresolved assembly 'Xamarin.iOS' reference.

There shouldn't be any references to Xamarin.iOS.dll in a .NET project.

Unfortunately my attempt to make the logs a bit more verbose was incorrect, so can you try this instead:

```xml
<MtouchExtraArgs>$(MtouchExtraArgs) -gcc_flags "-Xlinker -ld_classic" -v -v -v -v -v -v -v -v -v</MtouchExtraArgs>
```

That will hopefully show enough information for me to figure out where the reference to Xamarin.iOS comes from.

**Comment by cdavidyoung** [comment]
@rolfbjarne Here is the binlog.  Thanks for looking into this!

[msbuild.binlog.zip](https://github.com/xamarin/xamarin-macios/files/13079224/msbuild.binlog.zip)

**Comment by rolfbjarne** [comment]
This is really weird.

Your log shows this:

> Extracted third-party framework 'lame.framework' from '/Users/charlesyoung/.nuget/packages/laerdal.ffmpeg.audio/4.4.27/lib/netstandard2.0/Laerdal.FFmpeg.dll'

but the file `lib/netstandard2.0/Laerdal.FFmpeg.dll` from the NuGet [`Laerdal.FFmpeg.Audio`](https://www.nuget.org/packages/Laerdal.FFmpeg.Audio/4.4.27) does not contain `lame.framework` (while the file `lib/xamarinios10/Laerdal.FFmpeg.dll` from the same NuGet does).

It looks like your NuGet cache has gotten corrupted somehow.

Can you clear your NuGet cache (`rm -rf ~/.nuget/packages`), and then try building again?

If that doesn't work, please attach an updated binlog.

**Comment by cdavidyoung** [comment]
@rolfbjarne Good news.  It is now building and running on my mac!

Doing the "rm -rf ~/.nuget/packages" made me a little nervous so I started out with "rm -rf ~/.nuget/packages/laerdal.*", but this seems to have been sufficient to clear up the Nuget cache.

So for future reference, how did you know this was the problem?

**Comment by cdavidyoung** [comment]
@rolfbjarne Good news.  It is now building and running on my mac!

Doing the "rm -rf ~/.nuget/packages" made me a little nervous so I started out with "rm -rf ~/.nuget/packages/laerdal.*", but this seems to have been sufficient to clear up the Nuget cache.

So for future reference, how did you know this was the problem?

[msbuild.binlog.zip](https://github.com/xamarin/xamarin-macios/files/13114568/msbuild.binlog.zip)

**Comment by rolfbjarne** [comment]
That's great! I'll close this then.

> So for future reference, how did you know this was the problem?

This wasn't particularly easy, but:

1. The original error message mentions `lame.framework`

> but linking in object file (/Users/charlesyoung/Projects/jiffyLog2/obj/Debug/net7.0-maccatalyst/maccatalyst-x64/linker-cache/lame.framework/lame[x86_64][2](https://github.com/xamarin/xamarin-macios/issues/VbrTag.o)) built for 'iOS-simulator'

2. The `lame.framework` must come from somewhere, so I asked for the binlogs to try to figure this out.
3. In the binlogs I found all the package references, and knowing that `lame` is audio-related, `Laerdal.FFmpeg.Audio` sounded related, so I started investigating this NuGet.
4. It turns out that `Laerdal.FFmpeg.Audio` contains the `lame.framework`, but only the version of the NuGet that supports Xamarin.iOS. Your binlog showed that you weren't using that version (you're using the `netstandard2.0` version, which is correct for a .NET project).
5. I tested a Mac Catalyst project locally, and it built just fine, referencing the `netstandard2.0` version of `Laerdal.FFmpeg.Audio`, and nothing brought `lame.framework` into the build.
6. At this point the only theory I had was that somehow your NuGet cache got corrupted, where the `netstandard2.0` version of `Laerdal.FFmpeg.Audio` is actually the Xamarin.iOS version.
7. Clearing the cache worked, so that confirmed it :)

**Comment by cdavidyoung** [comment]
Thanks, @rolfbjarne.  I would have never figured that out.

Laerdal.FFmpeg.Audio used to work for both iOS and Android on my previous Xamarin version of this app.  When I converted it to Maui it stopped working for iOS and my intention is to eventually redo the Nuget locally to work with iOS again but I have not gotten around to it.  It is not critical.

Is there some best practice that I can follow to avoid something like this in the future?

**Comment by rolfbjarne** [comment]
> Is there some best practice that I can follow to avoid something like this in the future?

This seems like a weird one-off event, so no idea really.

**Comment by cdavidyoung** [comment]
@rolfbjarne Sorry to bother you again but while I am able to run my app it has strange behaviors associated with menu bar items and other things.  I decided to just try out some of the maui-samples to see how they behave.  There is some strange stuff going on.  For instance, in the ControlGallery Picker Demo the list of colors does not come up.  Does it work for you?  If you like, I can just start a new topic.

![image](https://github.com/xamarin/xamarin-macios/assets/937899/3fa041d4-e096-48fa-b5a1-73d065771e2d)

**Comment by rolfbjarne** [comment]
> @rolfbjarne Sorry to bother you again but while I am able to run my app it has strange behaviors associated with menu bar items and other things. I decided to just try out some of the maui-samples to see how they behave. There is some strange stuff going on. For instance, in the ControlGallery Picker Demo the list of colors does not come up. Does it work for you? If you like, I can just start a new topic.
> 
> ![image](https://user-images.githubusercontent.com/937899/278168178-3fa041d4-e096-48fa-b5a1-73d065771e2d.png)

Please file a new issue, this is unlikely to be related to the build errors.
